File: python/day5/part2.py
import re
import locale
import time

locale.setlocale(locale.LC_ALL, '')  # Use '' for auto, or force e.g. to 'en_US.UTF-8'


def determine_location(seed_input, mapping_dict):
    start_value = int(seed_input)

    for cat, range_list in mapping_dict.items():
        for mapping in range_list:
            mapping_source, mapping_destination, mapping_length = mapping
            mapping_source_end = mapping_source + mapping_length - 1
            if mapping_source <= start_value <= mapping_source_end:
                start_value += (mapping_destination - mapping_source)
                break

    return start_value


if __name__ == "__main__":
    f = open("input.txt", "r")

    seeds: [int] = []
    mappings: dict[str, list[list[int]]] = {}

    category = None
    for x in f:
        sanitized = re.sub(' +', ' ', x.replace('\n', '')).strip()

        if sanitized.startswith('seeds:'):
            seeds.extend(sanitized.replace('seeds: ', '').split(' '))
            continue

        if sanitized.strip() == '':
            continue

        if not sanitized[0].isnumeric():
            category = sanitized.replace(' map:', '')
            mappings[category] = []
        else:
            destination = int(sanitized.split(' ')[0])
            source = int(sanitized.split(' ')[1])
            length = int(sanitized.split(' ')[2])

            mappings[category].append([source, destination, length])

    print_debug_output = False

    total_amount_of_seeds = 0
    i = 0
    while i < len(seeds):
        total_amount_of_seeds += len(range(int(seeds[i]), int(seeds[i]) + int(seeds[i + 1]), 1))
        i += 2

    start = time.time()
    i = 0
    cnt = 0
    lowest = None
    last_cat = None
    last_print_at = None

    while i < len(seeds):
        seed_range = range(int(seeds[i]), int(seeds[i]) + int(seeds[i + 1]), 1)
        for seed in seed_range:
            cnt += 1
            if cnt % 1300000 == 0:
                print('Progress {:0.3f} ({:n}/{:n}, took {:0.0f}s)'.format(
                    (cnt / total_amount_of_seeds) * 100, cnt, total_amount_of_seeds, time.time() - start))
            source_value = determine_location(int(seed), mappings)

            if lowest is None:
                lowest = source_value
            elif lowest > source_value:
                lowest = source_value

        i += 2

    print('\n Lowest ' + ' = ' + str(lowest))

File: java/day5/Part2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.LongStream;

public class Part2 {
    public static void main(String[] args) throws IOException {
        var seeds = new ArrayList<Long>();
        // Linked is essential, since the almanac order is important
        var mappings = new LinkedHashMap<String, List<List<Long>>>();

        String category = null;
        try (BufferedReader br = new BufferedReader(new FileReader("day5/input.txt"))) {
            String line;
            while ((line = br.readLine()) != null) {
                var sanitized = line.replaceAll("\\s+", " ");

                if (sanitized.startsWith("seeds: ")) {
                    seeds.addAll(
                            Arrays.stream(
                                    sanitized.replaceAll("seeds: ", "").split(" ")
                            ).map(Long::parseLong).toList());
                }

                if (sanitized.trim().isEmpty()) {
                    continue;
                }

                if (!Character.isDigit(sanitized.charAt(0))) {
                    category = sanitized.replace(" map:", "");
                    mappings.putIfAbsent(category, new ArrayList<>());
                } else {
                    List<Long> parts = Arrays.stream(sanitized.split(" ")).map(Long::parseLong).toList();
                    long[] orderedParts = new long[]{parts.get(1), parts.get(0), parts.get(2)};
                    mappings.get(category).add(Arrays.stream(orderedParts).boxed().toList());
                }
            }

            long start = System.nanoTime();
            var i = 0;
            AtomicLong lowest = new AtomicLong(Long.MAX_VALUE);

            while (i < seeds.size()) {
                LongStream.range(seeds.get(i), seeds.get(i) + seeds.get(i + 1)).parallel().forEach(
                        seed -> {
                            final long location = Part2.determineLocation(seed, mappings);
                            lowest.getAndUpdate((val) -> val == Long.MAX_VALUE || val > location ? location : val);
                        }
                );

                i += 2;
            }

            long endTime = System.nanoTime();
            double timeTakenInMillis = (endTime - start) / 1_000_000_000.0;
            double roundedTime = Math.round(timeTakenInMillis * 1000.0) / 1000.0;
            System.out.println("Lowest = " + lowest + " took " + roundedTime + "s");

        }
    }

    public static Long determineLocation(final long seed, HashMap<String, List<List<Long>>> mappings) {
        long startValue = seed;

        for (Map.Entry<String, List<List<Long>>> entry : mappings.entrySet()) {
            for (List<Long> mapping : entry.getValue()) {
                /* 0 = source, 1 = destination, 2 = length */
                var mappingSource = mapping.get(0);
                var mappingDestination = mapping.get(1);
                var mappingSourceEnd = mappingSource + mapping.get(2) - 1;

                if (startValue >= mappingSource && startValue <= mappingSourceEnd) {
                    startValue += (mappingDestination - mappingSource);
                    break;
                }
            }
        }

        return startValue;
    }
}